### PR TITLE
Implement soft delay extension

### PR DESCRIPTION
**Pull request by btasdelen**
Work in progress. Effort to implement soft delay extension for file format 1.5.0 compatibility.

TODO: 
-------
- [ ] Add `soft_delay` handling to  `addBlock()`.
- [ ] Implement `apply_soft_delay()`.
- [ ] Add `soft_delay` handling to `write()`.
- [ ] Implement `get_default_soft_delay_values()`
- [ ] Add tests for soft delay.
- [ ] Add example sequence.

### CODE REVIEW

**Comment by FrankZijlstra** [comment]
Regarding the failing sequence tests, you can set the environment variable SAVE_EXPECTED=1 and run pytest to locally rewrite the .seq files. This will automatically pass the write tests, but then allows you to do a diff with the previous .seq files (e.g. git diff, or make a copy of the .seq files first). Then if you manually verify that the change of IDs is expected, commit the new .seq files. I know manually checking this is a bit of a pain, but I think aiming for a exact match of the .seq files is a good standard, better safe than sorry.

File: src/pypulseq/Sequence/block.py [hunk 1]
```diff
@@ -150,9 +155,22 @@ def set_block(self, block_index: int, *args: SimpleNamespace) -> None:
...
+                if hasattr(event, 'id'):
+                    event_id = event.id
+                else:
+                    event_id = register_soft_delay_event(self, event)
+                ext = {'type': self.get_extension_type_ID('DELAYS'), 'ref': event_id}
+                extensions.append(ext)
+            else:
+                raise ValueError(f'Unknown event type {event.type} passed to set_block().')
         else:
             # Floating point number given as delay
```
Comment (fzimmermann89) [inline src/pypulseq/Sequence/block.py:168]: delay given as floating poiint number

File: src/pypulseq/Sequence/block.py [hunk 2]
```diff
@@ -40,11 +40,16 @@ def set_block(self, block_index: int, *args: SimpleNamespace) -> None:
...
         If a gradient that doesn't end at zero is not aligned to the block boundary.
+        If multiple soft_delay extensions are used in a block.
+        If a soft delay extension is used in a block of zero duration.
+        If a soft delay extension is used in a block containing conventional events.
     """
     events = block_to_events(*args)
     new_block = np.zeros(7, dtype=np.int32)
     duration = 0
 
+    required_duration = None
```
Comment (fzimmermann89) [inline src/pypulseq/Sequence/block.py:51]:
```suggestion
    required_duration:float|None = None
```

File: src/pypulseq/make_soft_delay.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+def make_soft_delay(numID: int, hint: str, offset: float = 0.0, factor: float = 1.0) -> SimpleNamespace:
+    """
+    Creates a soft delay extension event.
+
+    Create a soft delay event, that can be used in combination with an empty (pure delay) block
+    e.g. to adjust TE, TR or other delays. The soft delay extension acts by rewriting the block duration
+    based on the user input (to the interpreter) according to the equation dur=input/factor+offset.
+    Required parameters are 'numeric ID' and 'string hint'. Optional parameter 'factor' can be either
+    positive and negative. Optional parameter 'offset' given in seconds can also be either positive and
+    negative. The 'hint' parameter is expected to be for identical 'numID'.
```
Comment (fzimmermann89) [inline src/pypulseq/make_soft_delay.py:13]: The 'hint' parameter is expected to be _identical_ (?) for identical 'numID'.

**Review by fzimmermann89 — COMMENTED** [review]
I had a quick look -- and noticed i need more time too understand the logic and play around with the code....

**Comment by fzimmermann89** [comment]
regarding check timings: for it to make sense, it would need to get the information about the values that should be used during the check, correct? maybe a dict of values?
 and check_timings should warn that the timings are only checked for this particular setting of delay values, other settings in the scanner UI can break.

File: src/pypulseq/Sequence/sequence.py [hunk 4]
```diff
@@ -1433,6 +1487,204 @@ def set_extension_string_ID(self, extension_str: str, extension_id: int) -> None
...
+        """
+
+        # TODO: check if this works
+        # Go through all the blocks and update durations, at the same time checking the consistency of the soft delays
+        sd_str2numID = {}
+        sd_numID2hint = {}
+        sd_warns = {}
+        for block_counters in self.block_events:
+            block = self.get_block(block_counters)
+            if hasattr(block, 'soft_delay') and block.soft_delay is not None:
```
Comment (FrankZijlstra) [inline src/pypulseq/Sequence/sequence.py:1525]: Add `soft_delay` to `attrs`  at top of `get_block` to always set it to `None`. Then remove the `hasattr` check here. Same in `get_default_soft_delay_values` below.

File: src/pypulseq/make_soft_delay.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    See Also
+    --------
+    - `pypulseq.Sequence.sequence.Sequence.add_block()`
+    - `pypulseq.Sequence.sequence.Sequence.apply_soft_delay()`
+
+    Parameters
+    ----------
+    numID : int
```
Comment (FrankZijlstra) [inline src/pypulseq/make_soft_delay.py:22]: I don't understand the reason the `numID` parameter exists and is not just assigned automatically, as is done for other events. If the `hint` needs to be identical for identical `numID`, then the delay can be uniquely addressed with just the `hint`.

**Comment by btasdelen** [comment]
@FrankZijlstra @fzimmermann89 Many thanks for the comments. It looks like #288 takes this PR and has many changes on top of this. I am not sure how to proceed with this without duplicating effort and creating some conflicts.

It looks like there are some changes to soft delay part, but most changes on #288 are for other features. If @mcencini wants to adress these feedback on there I can close this PR.

File: src/pypulseq/make_soft_delay.py (hunk 5 above)
Comment (btasdelen) [reply src/pypulseq/make_soft_delay.py:22]: I think the main purpose of the `numID` is letting the user choose which delay box they want to use for a specific delay. For example, if they want to put TE first then TR, they can do so by explicitly assigning IDs 0 and 1, to them, regardless of the order they are added into the sequence.

But, I agree with you. It creates confusion for little gain. Also, I tried setting `numID` to 2 and did not assign anything to 0 and 1, and interpreter shows 3 delays anyway since it has to create them sequentially. It gracefully handled empty events, but still.

I think we can deviate from the Matlab here.

File: src/pypulseq/Sequence/block.py [hunk 6]
```diff
@@ -264,11 +296,18 @@ def set_block(self, block_index: int, *args: SimpleNamespace) -> None:
...
+        duration = required_duration
+
     self.block_events[block_index] = new_block
     self.block_durations[block_index] = float(duration)
 
 
-def get_block(self, block_index: int) -> SimpleNamespace:
+def get_block(self, block_index: int, add_ids: bool = False) -> SimpleNamespace:
     """
     Returns PyPulseq block at `block_index` position in `self.block_events`.
```
[the remark this answers is not in the document]
Comment (btasdelen) [reply src/pypulseq/Sequence/block.py:312]: Good catch. I could not figure out the exact purpose yet, it looks like when true, it assigns an `id` member to all extensions. The only usage of this is I found in `Transform` functions: 
https://github.com/pulseq/pulseq/blob/43c281e5d99ca0bdd99b1334632ad2762527a2ef/matlab/%2Bmr/transform.m#L106
and
https://github.com/pulseq/pulseq/blob/43c281e5d99ca0bdd99b1334632ad2762527a2ef/matlab/%2Bmr/%40TransformFOV/TransformFOV.m#L359

For now, I will remove it since it is not relevant to soft delays. We can decide if we need it later on.

File: examples/scripts/write_gre_label_softdelay.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+                channel='y',
+                area=phase_areas[i],
+                duration=pp.calc_duration(gx_pre),
+                system=system,
+            )
+            seq.add_block(gx_pre, gy_pre, gz_reph)
+            # seq.add_block(pp.make_delay(delay_TE))
+            seq.add_block(
+                10e-6,  # Add a small delay as we can't have soft_delay in an 0 duration block.
+                pp.make_soft_delay(numID=0, hint='TE', offset=-min_TE, factor=1.0),
```
[the remark this answers is not in the document]
Comment (btasdelen) [reply examples/scripts/write_gre_label_softdelay.py]: I agree. For now, I removed the `required_duration`, now `default_duration` is an input to `make_soft_delay` with a default value of 10-e5.

> On a sidenote, I really think the float input as an option to add_block should not exist in the first place.

I looked into this, but I felt like removing that will cause bunch of other issues to solve: https://github.com/imr-framework/pypulseq/commit/99bbffd058d1284124a58a4c68b591205b147328

File: src/pypulseq/Sequence/block.py (hunk 2 above)
Comment (btasdelen) [reply src/pypulseq/Sequence/block.py]: `required_duration` is removed.

File: src/pypulseq/Sequence/block.py [hunk 8]
```diff
@@ -13,7 +13,7 @@
 from pypulseq.utils.tracing import trace_enabled
 
 
-def set_block(self, block_index: int, *args: SimpleNamespace) -> None:
+def set_block(self, block_index: int, *args: Union[SimpleNamespace, float]) -> None:
```
[the remark this answers is not in the document]
Comment (btasdelen) [reply src/pypulseq/Sequence/block.py:16]: Done.

**Comment by FrankZijlstra** [comment]
I think it would be okay to merge it with a few minor changes:
- Add write_gre_label_softdelay to the example sequence tests in `test_sequence`. This checks basic things like writing and reading the sequence, which is good to test the soft delay extension blocks in the sequence file.
- Fix `get_default_soft_delay_values` so it does not contain the error report stuff. I do not fully understand the checks being done, but it seems to me it should be in `check_timing` (or maybe it can be checked immediately when the block gets added?). These can be left as placeholders wherever it makes most sense to put these checks. Note that `check_timing` automatically gets called before writing a sequence, so it is the most safe for any sequence checks that would breaks things on the scanner.

**Review by schuenke — APPROVED** [review]
I refactored the code quite a lot trying to take all previous comments into account. You can see a summary of all my changes using the GitHub comparison functionality:

https://github.com/imr-framework/pypulseq/compare/0374a60...40f17f5

The most important changes are:

- **Simplified API**: Made numID optional with automatic assignment based on hint order
- **Complete documentation**: Added comprehensive docstrings with examples and mathematical explanations  
- **Robust validation**: Improved error messages and parameter validation (including all whitespace types)
- **Comprehensive testing**: Added 14 test functions covering edge cases and boundary conditions
- **Better examples**: Created simple beginner example + improved complex example documentation
- **Code cleanup**: Removed broken functions, improved duration handling, cleaner float API separation

Would be great if @btasdelen @FrankZijlstra @mcencini and maybe @fzimmermann89 could have a look at it again. 

I am not sure about the soft_delay_simple_example, because it doesn't really fit into the available examples, but still thought it might be helpful. The documentation would probably be a good place for this, but IMO the docu is not the best atm and thus not really used by anyone...
